File: go/languages.go
// languages.go
//
// WritingSystem, LanguageTag, DetectionResult, KnownLanguages,
// ILanguageDetector, ILanguageRegistry.

package circleai

import "context"

// ---------------------------------------------------------------------------
// WritingSystem
// ---------------------------------------------------------------------------

// WritingSystem is the script used by a language.
type WritingSystem int

const (
	WritingSystemLatin      WritingSystem = iota
	WritingSystemArabic
	WritingSystemEthiopic
	WritingSystemGeez       // alias for Ethiopic in the C# original; kept for completeness
	WritingSystemDevanagari
	WritingSystemHan
	WritingSystemCyrillic
	WritingSystemHebrew
	WritingSystemGreek
	WritingSystemOther
)

// ---------------------------------------------------------------------------
// LanguageTag
// ---------------------------------------------------------------------------

// LanguageTag is a BCP-47 language tag enriched with display metadata.
type LanguageTag struct {
	// BcpTag is the IETF BCP-47 language tag (e.g. "en", "zu", "ar").
	BcpTag string

	// EnglishName is the English display name of the language.
	EnglishName string

	// NativeName is the name of the language in that language.
	NativeName string

	// WritingSystem is the primary script used by this language.
	WritingSystem WritingSystem

	// IsRtl indicates whether the language is written right-to-left.
	IsRtl bool

	// PrimaryRegion is the ISO 3166-1 alpha-2 region code where this
	// language is primarily spoken (e.g. "ZA", "NG").
	PrimaryRegion string
}

// LanguageTagUnknown is the sentinel returned when detection fails.
var LanguageTagUnknown = LanguageTag{
	BcpTag:        "und",
	EnglishName:   "Unknown",
	NativeName:    "Unknown",
	WritingSystem: WritingSystemLatin,
	IsRtl:         false,
	PrimaryRegion: "",
}

// ---------------------------------------------------------------------------
// DetectionResult
// ---------------------------------------------------------------------------

// DetectionResult is the result of a language detection operation.
type DetectionResult struct {
	// Language is the detected language tag.
	Language LanguageTag

	// Confidence is the detector's confidence in [0, 1].
	Confidence float32

	// IsReliable indicates whether the result is considered reliable.
	IsReliable bool
}

// ---------------------------------------------------------------------------
// ScriptNormalisationResult
// ---------------------------------------------------------------------------

// ScriptNormalisationResult is the result of script normalisation.
type ScriptNormalisationResult struct {
	// Input is the original text.
	Input string

	// Normalised is the text after script normalisation.
	Normalised string

	// DetectedLanguage is the language detected during normalisation.
	DetectedLanguage LanguageTag
}

// ---------------------------------------------------------------------------
// KnownLanguages
// ---------------------------------------------------------------------------

// Africa
var (
	LangIsiZulu  = LanguageTag{BcpTag: "zu", EnglishName: "isiZulu", NativeName: "isiZulu", WritingSystem: WritingSystemLatin, IsRtl: false, PrimaryRegion: "ZA"}
	LangSesotho  = LanguageTag{BcpTag: "st", EnglishName: "Sesotho", NativeName: "Sesotho", WritingSystem: WritingSystemLatin, IsRtl: false, PrimaryRegion: "ZA"}
	LangAfrikaans = LanguageTag{BcpTag: "af", EnglishName: "Afrikaans", NativeName: "Afrikaans", WritingSystem: WritingSystemLatin, IsRtl: false, PrimaryRegion: "ZA"}
	LangSwahili  = LanguageTag{BcpTag: "sw", EnglishName: "Swahili", NativeName: "Kiswahili", WritingSystem: WritingSystemLatin, IsRtl: false, PrimaryRegion: "KE"}
	LangHausa    = LanguageTag{BcpTag: "ha", EnglishName: "Hausa", NativeName: "Hausa", WritingSystem: WritingSystemLatin, IsRtl: false, PrimaryRegion: "NG"}
	LangAmharic  = LanguageTag{BcpTag: "am", EnglishName: "Amharic", NativeName: "አማርኛ", WritingSystem: WritingSystemEthiopic, IsRtl: false, PrimaryRegion: "ET"}
	LangYoruba   = LanguageTag{BcpTag: "yo", EnglishName: "Yoruba", NativeName: "Yorùbá", WritingSystem: WritingSystemLatin, IsRtl: false, PrimaryRegion: "NG"}
	LangIgbo     = LanguageTag{BcpTag: "ig", EnglishName: "Igbo", NativeName: "Igbo", WritingSystem: WritingSystemLatin, IsRtl: false, PrimaryRegion: "NG"}
	LangXhosa    = LanguageTag{BcpTag: "xh", EnglishName: "isiXhosa", NativeName: "isiXhosa", WritingSystem: WritingSystemLatin, IsRtl: false, PrimaryRegion: "ZA"}
	LangSepedi   = LanguageTag{BcpTag: "nso", EnglishName: "Sepedi", NativeName: "Sepedi", WritingSystem: WritingSystemLatin, IsRtl: false, PrimaryRegion: "ZA"}
	LangSetswana = LanguageTag{BcpTag: "tn", EnglishName: "Setswana", NativeName: "Setswana", WritingSystem: WritingSystemLatin, IsRtl: false, PrimaryRegion: "ZA"}
	LangSomali   = LanguageTag{BcpTag: "so", EnglishName: "Somali", NativeName: "Soomaali", WritingSystem: WritingSystemLatin, IsRtl: false, PrimaryRegion: "SO"}
	LangOromo    = LanguageTag{BcpTag: "om", EnglishName: "Oromo", NativeName: "Afaan Oromoo", WritingSystem: WritingSystemLatin, IsRtl: false, PrimaryRegion: "ET"}
)

// Middle East & North Africa
var (
	LangArabic = LanguageTag{BcpTag: "ar", EnglishName: "Arabic", NativeName: "العربية", WritingSystem: WritingSystemArabic, IsRtl: true, PrimaryRegion: "SA"}
)

// Europe & Americas
var (
	LangEnglish    = LanguageTag{BcpTag: "en", EnglishName: "English", NativeName: "English", WritingSystem: WritingSystemLatin, IsRtl: false, PrimaryRegion: "GB"}
	LangPortuguese = LanguageTag{BcpTag: "pt", EnglishName: "Portuguese", NativeName: "Português", WritingSystem: WritingSystemLatin, IsRtl: false, PrimaryRegion: "PT"}
	LangFrench     = LanguageTag{BcpTag: "fr", EnglishName: "French", NativeName: "Français", WritingSystem: WritingSystemLatin, IsRtl: false, PrimaryRegion: "FR"}
	LangSpanish    = LanguageTag{BcpTag: "es", EnglishName: "Spanish", NativeName: "Español", WritingSystem: WritingSystemLatin, IsRtl: false, PrimaryRegion: "ES"}
)

// Asia
var (
	LangMandarin = LanguageTag{BcpTag: "zh", EnglishName: "Mandarin", NativeName: "中文", WritingSystem: WritingSystemHan, IsRtl: false, PrimaryRegion: "CN"}
	LangHindi    = LanguageTag{BcpTag: "hi", EnglishName: "Hindi", NativeName: "हिन्दी", WritingSystem: WritingSystemDevanagari, IsRtl: false, PrimaryRegion: "IN"}
)

// KnownLanguagesAll holds all 20 languages shipped with Circle AI,
// in declaration order.
var KnownLanguagesAll = []LanguageTag{
	LangIsiZulu, LangSesotho, LangAfrikaans, LangSwahili, LangHausa, LangAmharic,
	LangYoruba, LangIgbo, LangXhosa, LangSepedi, LangSetswana, LangSomali, LangOromo,
	LangArabic,
	LangEnglish, LangPortuguese, LangFrench, LangSpanish,
	LangMandarin, LangHindi,
}

// ---------------------------------------------------------------------------
// ILanguageDetector
// ---------------------------------------------------------------------------

// ILanguageDetector detects the BCP-47 language of a piece of text.
type ILanguageDetector interface {
	// Detect detects the most likely language. Returns LanguageTagUnknown
	// with Confidence=0 when detection fails.
	Detect(ctx context.Context, text string) (DetectionResult, error)

	// DetectMultiple returns up to maxResults candidates ranked by confidence.
	DetectMultiple(ctx context.Context, text string, maxResults int) ([]DetectionResult, error)
}

// ---------------------------------------------------------------------------
// ILanguageRegistry
// ---------------------------------------------------------------------------

// ILanguageRegistry is the registry of all BCP-47 language tags that
// Circle AI understands.
type ILanguageRegistry interface {
	// GetByBcpTag returns the LanguageTag for the given BCP-47 tag,
	// or nil if not found.
	GetByBcpTag(bcpTag string) *LanguageTag

	// GetAll returns all language tags in the registry.
	GetAll() []LanguageTag

	// GetForRegion returns all language tags whose primary region matches
	// the given ISO 3166-1 alpha-2 region code.
	GetForRegion(isoRegion string) []LanguageTag

	// IsSupported reports whether the given BCP-47 tag is in the registry.
	IsSupported(bcpTag string) bool
}

File: c/src/languages.c
/*
 * languages.c — KnownLanguages static registry (20 entries).
 *
 * All entries match fixtures/language_tags.json exactly:
 *   - english_name matches the "englishName" field
 *   - native_name  matches the "nativeName"  field (UTF-8 encoded)
 *   - primary_region matches the fixture (en -> "GB", pt -> "PT")
 *   - writing_system uses the enum order from languages.h:
 *       LATIN=0, ARABIC=1, ETHIOPIC=2, HANZI=3, DEVANAGARI=4
 *
 * Pure C11, no OS-specific headers.
 */

#include "circle_ai/languages.h"
#include <string.h>

/*
 * UTF-8 byte sequences for non-ASCII native names:
 *
 *   Amharic   "አማርኛ"  : \xe1\x8a\xa0\xe1\x88\x9b\xe1\x88\xad\xe1\x8a\x9b
 *   Yoruba    "Yorùbá" : Yor\xc3\xb9b\xc3\xa1
 *   Arabic    "العربية": \xd8\xa7\xd9\x84\xd8\xb9\xd8\xb1\xd8\xa8\xd9\x8a\xd8\xa9
 *   Mandarin  "中文"   : \xe4\xb8\xad\xe6\x96\x87
 *   Hindi     "हिन्दी"  : \xe0\xa4\xb9\xe0\xa4\xbf\xe0\xa4\xa8\xe0\xa5\x8d\xe0\xa4\xa6\xe0\xa5\x80
 *   Portuguese "Português": Portugu\xc3\xaas
 *   French    "Français" : Fran\xc3\xa7ais
 *   Spanish   "Español"  : Espa\xc3\xb1ol
 */

static const ca_language_tag_t LANGUAGES[CA_LANGUAGE_COUNT] = {
    /* 0 */ { "zu",  "isiZulu",     "isiZulu",
              CA_WS_LATIN,      0, "ZA" },
    /* 1 */ { "st",  "Sesotho",     "Sesotho",
              CA_WS_LATIN,      0, "ZA" },
    /* 2 */ { "af",  "Afrikaans",   "Afrikaans",
              CA_WS_LATIN,      0, "ZA" },
    /* 3 */ { "sw",  "Swahili",     "Kiswahili",
              CA_WS_LATIN,      0, "KE" },
    /* 4 */ { "ha",  "Hausa",       "Hausa",
              CA_WS_LATIN,      0, "NG" },
    /* 5 */ { "am",  "Amharic",
              "\xe1\x8a\xa0\xe1\x88\x9b\xe1\x88\xad\xe1\x8a\x9b",
              CA_WS_ETHIOPIC,   0, "ET" },
    /* 6 */ { "yo",  "Yoruba",
              "Yor\xc3\xb9\x62\xc3\xa1",   /* Yorùbá */
              CA_WS_LATIN,      0, "NG" },
    /* 7 */ { "ig",  "Igbo",        "Igbo",
              CA_WS_LATIN,      0, "NG" },
    /* 8 */ { "xh",  "isiXhosa",    "isiXhosa",
              CA_WS_LATIN,      0, "ZA" },
    /* 9 */ { "nso", "Sepedi",      "Sepedi",
              CA_WS_LATIN,      0, "ZA" },
    /*10 */ { "tn",  "Setswana",    "Setswana",
              CA_WS_LATIN,      0, "ZA" },
    /*11 */ { "so",  "Somali",      "Soomaali",
              CA_WS_LATIN,      0, "SO" },
    /*12 */ { "om",  "Oromo",       "Afaan Oromoo",
              CA_WS_LATIN,      0, "ET" },
    /*13 */ { "ar",  "Arabic",
              "\xd8\xa7\xd9\x84\xd8\xb9\xd8\xb1\xd8\xa8\xd9\x8a\xd8\xa9",
              CA_WS_ARABIC,     1, "SA" },
    /*14 */ { "en",  "English",     "English",
              CA_WS_LATIN,      0, "GB" },
    /*15 */ { "pt",  "Portuguese",
              "Portugu\xc3\xaas",           /* Português */
              CA_WS_LATIN,      0, "PT" },
    /*16 */ { "fr",  "French",
              "Fran\xc3\xa7\x61\x69\x73",  /* Français  */
              CA_WS_LATIN,      0, "FR" },
    /*17 */ { "es",  "Spanish",
              "Espa\xc3\xb1\x6f\x6c",      /* Español   */
              CA_WS_LATIN,      0, "ES" },
    /*18 */ { "zh",  "Mandarin",
              "\xe4\xb8\xad\xe6\x96\x87",  /* 中文      */
              CA_WS_HANZI,      0, "CN" },
    /*19 */ { "hi",  "Hindi",
              "\xe0\xa4\xb9\xe0\xa4\xbf\xe0\xa4\xa8\xe0\xa5\x8d\xe0\xa4\xa6\xe0\xa5\x80",
              CA_WS_DEVANAGARI, 0, "IN" },
};

const ca_language_tag_t *ca_known_languages(void) {
    return LANGUAGES;
}

int ca_language_count(void) {
    return CA_LANGUAGE_COUNT;
}

const ca_language_tag_t *ca_find_language(const char *bcp_tag) {
    for (int i = 0; i < CA_LANGUAGE_COUNT; i++) {
        if (strcmp(LANGUAGES[i].bcp_tag, bcp_tag) == 0) {
            return &LANGUAGES[i];
        }
    }
    return NULL;
}
